### DISCUSSION TITLE
Cannot install ruby 3.2.2 with rbenv on macOS 13.5

### DISCUSSION THREAD

**Discussion by Dante1349** [rbenv/ruby-build#2293] [Build failures] [answered]
When i try to run 

```
rbenv install 3.2.2
``` 

it fails with this output:

```
-c -o crypto/evp/libcrypto-lib-bio_b64.o crypto/evp/bio_b64.c
make[1]: *** [crypto/err/libcrypto-lib-err_prn.o] Killed: 9
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [crypto/engine/libcrypto-lib-tb_rsa.o] Killed: 9
make[1]: *** Deleting file `crypto/engine/libcrypto-lib-tb_rsa.o'
/bin/sh: line 1: 81757 Killed: 9               mv crypto/engine/libcrypto-lib-tb_pkmeth.d.tmp crypto/engine/libcrypto-lib-tb_pkmeth.d
make[1]: *** [crypto/engine/libcrypto-lib-tb_pkmeth.o] Error 137
make[1]: *** [crypto/ess/libcrypto-lib-ess_err.o] Killed: 9
make[1]: *** [crypto/err/libcrypto-lib-err_all_legacy.o] Killed: 9
make[1]: *** Deleting file `crypto/err/libcrypto-lib-err_all_legacy.o'
make[1]: *** [crypto/ess/libcrypto-lib-ess_asn1.o] Killed: 9
make: *** [build_sw] Error 2

BUILD FAILED (macOS 13.5 using ruby-build 20231025)
```

i tried running it like this but also had no luck:
```
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 3.2.2
```
thats the output i get then:
```
To follow progress, use 'tail -f /var/folders/1w/k8rgm53s7t33qkby3xscbj_80000gq/T/ruby-build.20231102161952.87732.log' or pass --verbose
Downloading ruby-3.2.2.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.2/ruby-3.2.2.tar.gz
Installing ruby-3.2.2...
ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 13.5 using ruby-build 20231025)

Inspect or clean up the working tree at /var/folders/1w/k8rgm53s7t33qkby3xscbj_80000gq/T/ruby-build.20231102161952.87732.t3ReDJ
Results logged to /var/folders/1w/k8rgm53s7t33qkby3xscbj_80000gq/T/ruby-build.20231102161952.87732.log

Last 10 log lines:
compiling option.c
linking shared-object strscan.bundle
compiling unixsocket.c
compiling ifaddr.c
linking shared-object stringio.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
linking shared-object socket.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
make: *** [build-ext] Error 2

```

**Comment by mislav** [comment]
> i tried running it like this but also had no luck:
> 
> ```
> RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)" rbenv install 3.2.2
> ```

Hi, first a minor note: `openssl@1.1` from Homebrew is now deprecated because OpenSSL 1.1 isn't maintained anymore. You can try using `openssl@3` instead:
```
brew install openssl@3
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)" rbenv install 3.2.2
```

Second, the "last 10 log lines" output rarely includes the actual build failure, in my experience. Please look into the log file indicated at the `Results logged to <FILE>` line and search for keywords like "error" or "failure". If you isolate something that looks like a proper build failure, you can paste it here (if it's relatively short) or as a Gist (if the output is really long, like a whole build log file). Thanks!

**Comment by Dante1349** [comment]
Thank you, i will try to look up in the logfile, but openssl@3 is also not working since this is my default installation in homebrew. I will provide more logs then.

**Comment by Dante1349** [comment]
when i run it with openssl@3 i get this in the logfile:
```
compiling ripper.c
compiling init.c
make[2]: *** [readline.o] Killed: 9
make[2]: *** [../../.ext/common/psych/coder.rb] Killed: 9
clang: make[1]: *** [ext/readline/all] Error 2
error: unable to execute command: Killed: 9
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [ext/psych/all] Error 2
make[1]: *** [ext/stringio/all] Killed: 9
compiling ossl_hmac.c
compiling constants.c
clang: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/1w/k8rgm53s7t33qkby3xscbj_80000gq/T/limits-9b34dc.c
clang: note: diagnostic msg: /var/folders/1w/k8rgm53s7t33qkby3xscbj_80000gq/T/limits-9b34dc.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /Users/$USER/Library/Logs/DiagnosticReports/clang_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg:

********************
make[2]: *** [limits.o] Error 1
make[1]: *** [ext/rbconfig/sizeof/all] Error 2
compiling ossl_kdf.c
compiling basicsocket.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling strscan.c
compiling ossl_ns_spki.c
compiling ossl_ocsp.c
installing default libraries
checking ../.././parse.y and ../.././ext/ripper/eventids2.c
compiling ossl_pkcs12.c
ripper.c:10310:9: warning: variable 'yynerrs' set but not used [-Wunused-but-set-variable]
    int yynerrs = 0;
        ^
compiling syslog.c
compiling ossl_pkcs7.c
compiling socket.c
compiling ipsocket.c
compiling ossl_pkey.c
compiling tcpsocket.c
installing default ripper libraries
linking shared-object strscan.bundle
installing default syslog libraries
compiling tcpserver.c
compiling ossl_pkey_dh.c
linking shared-object syslog.bundle
compiling udpsocket.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling ossl_pkey_dsa.c
ossl_pkey.c:326:22: warning: implicit truncation from 'int' to a one-bit wide bit-field changes value from 1 to -1 [-Wsingle-bit-bitfield-constant-conversion]
    arg->interrupted = 1;
                     ^ ~
compiling sockssocket.c
compiling ossl_pkey_ec.c
1 warning generated.
compiling unixsocket.c
compiling unixserver.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling option.c
compiling ossl_pkey_rsa.c
compiling ossl_rand.c
compiling ossl_ssl.c
compiling ancdata.c
ossl_pkey_ec.c:489:5: warning: assigning to 'EC_KEY *' (aka 'struct ec_key_st *') from 'const struct ec_key_st *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    GetEC(self, ec);
    ^~~~~~~~~~~~~~~
ossl_pkey_ec.c:24:11: note: expanded from macro 'GetEC'
    (key) = EVP_PKEY_get0_EC_KEY(_pkey); \
          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~
compiling ossl_ssl_session.c
1 warning generated.
compiling ossl_x509.c
1 warning generated.
compiling raddrinfo.c
compiling ossl_ts.c
compiling ossl_x509attr.c
compiling ifaddr.c
linking shared-object ripper.bundle
compiling ossl_x509cert.c
compiling ossl_x509crl.c
compiling ossl_x509ext.c
compiling ossl_x509req.c
compiling ossl_x509revoked.c
compiling ossl_x509name.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling ossl_x509store.c
linking shared-object socket.bundle
installing default openssl libraries
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
linking shared-object openssl.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
make: *** [build-ext] Error 2
```

the crash log thats mentioned here does not exist. There is no clang_* file. Maybe there is a permission problem.

**Reply by mislav** [reply]
This looks like some general error with the clang compiler on your system, but I don't know what causes it.

What is the output of `xcode-select -p`?
```
$ xcode-select -p
/Library/Developer/CommandLineTools
```

**Reply by Dante1349** [reply]
`/Applications/Xcode.app/Contents/Developer` is the output on my machine

**Reply by Dante1349** [reply]
its a clean mac install btw. on an M2 ARM core. Im trying to setup an ansible installer for our developer clients and i don't want to use the system ruby

**Reply by mislav** [reply]
@Dante1349 Does anything change if you do `sudo xcode-select --reset` and then try again?

I don't know what causes clang to error out, but I doubt that it is related to any `openssl` version. It seems that the compiler is busted somehow. Strange, since it's a clean Mac install! I cannot reproduce on my M1 machine, but I do not have access to an M2.

You could try re-installing Xcode (or installing a different version) and trying again.

**Reply by Dante1349** [reply]
@mislav I think its pretty much the same result:

````
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling qsort.c
bigdecimal.c:7161:18: warning: variable 'prec' set but not used [-Wunused-but-set-variable]
    SIGNED_VALUE prec;
                 ^
compiling digest.c
installing default coverage libraries
compiling date_parse.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
linking shared-object coverage.bundle
compiling date_strftime.c
compiling rb_interned_str.c
compiling rb_str_dup.c
compiling escape.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling set_len.c
1 warning generated.
compiling etc.c
compiling date_strptime.c
compiling fcntl.c
linking shared-object bigdecimal.bundle
compiling closure.c
linking shared-object digest.bundle
compiling console.c
linking shared-object erb/escape.bundle
compiling nonblock.c
linking shared-object -test-/string.bundle
linking shared-object etc.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
linking shared-object fcntl.bundle
compiling wait.c
compiling conversions.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
installing digest libraries
compiling fiddle.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
installing default date_core libraries
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
linking shared-object io/nonblock.bundle
linking shared-object date_core.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling function.c
linking shared-object io/wait.bundle
compiling handle.c
linking shared-object io/console.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling monitor.c
installing default console libraries
compiling nkf.c
compiling memory_view.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling pinned.c
compiling pointer.c
compiling object_tracing.c
compiling objspace.c
linking shared-object monitor.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling objspace_dump.c
compiling openssl_missing.c
compiling pathname.c
compiling psych.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling pty.c
installing default fiddle libraries
installing default pathname libraries
compiling cparse.c
compiling psych_emitter.c
linking shared-object nkf.bundle
compiling ossl.c
installing default pty libraries
linking shared-object pathname.bundle
compiling sizes.c
compiling psych_parser.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
linking shared-object pty.bundle
linking shared-object racc/cparse.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling readline.c
linking shared-object objspace.bundle
compiling ossl_asn1.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling limits.c
compiling psych_to_ruby.c
compiling ossl_bio.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling psych_yaml_tree.c
linking shared-object fiddle.bundle
installing default objspace libraries
compiling ripper.c
compiling ossl_bn.c
linking shared-object readline.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
linking shared-object rbconfig/sizeof.bundle
make[2]: *** [ossl_bn.o] Killed: 9
make[2]: *** [../../../.ext/arm64-darwin23/rbconfig/sizeof.bundle] Killed: 9
make[1]: *** [ext/openssl/all] Error 2
make[2]: *** [psych_yaml_tree.o] Killed: 9
make[2]: *** Waiting for unfinished jobs....
make[1]: *** Waiting for unfinished jobs....
clang: error: unable to execute command: Killed: 9
clang: error: linker command failed due to signal (use -v to see invocation)
make[2]: *** [../../.ext/arm64-darwin23/fiddle.bundle] Error 1
make[1]: *** [ext/fiddle/all] Error 2
make[2]: *** [../../.ext/common/objspace/trace.rb] Killed: 9
make[1]: *** [ext/objspace/all] Error 2
make[2]: *** [../../.ext/.timestamp/.RUBYCOMMONDIR.-.ripper.time] Killed: 9
make[2]: *** Waiting for unfinished jobs....
checking ../.././parse.y and ../.././ext/ripper/eventids2.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
make[1]: *** [ext/psych/all] Error 2
compiling init.c
compiling basicsocket.c
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
compiling sockssocket.c
compiling socket.c
compiling udpsocket.c
compiling unixsocket.c
compiling tcpserver.c
ripper.c:10310:9: warning: variable 'yynerrs' set but not used [-Wunused-but-set-variable]
    int yynerrs = 0;
        ^
compiling unixserver.c
compiling constants.c
compiling ipsocket.c
compiling option.c
compiling tcpsocket.c
compiling ancdata.c
compiling raddrinfo.c
compiling ifaddr.c
linking shared-object socket.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
1 warning generated.
make[1]: *** [ext/ripper/all] Error 2
make: *** [build-ext] Error 2
````

I also updated to macos14.1 but no changes.

**Reply by Dante1349** [reply]
i ran: `RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)" rbenv install 3.2.2` it didn't work either without the opts

**Reply by mislav** [reply]
Weird. The internet is full of reports their system killing `clang` processes by sending it the KILL signal (9), but none are really conclusive as to what the root cause is. https://developer.apple.com/forums/thread/701242

Sometimes people saw improvements after system restart, the other times they reinstalled either Xcode or Command Line Tools to work around this.

Other things you can do is set the TMPDIR environment variable to another location on disk where ruby-build compilation steps should happen, in case something is wrong with the default TMPDIR location, or set `MAKE_OPTS='-j2'` to reduce the number of parallel build processes (the default parallelism number is the number of CPUs on your machine, which can be around 8).

**Reply by Dante1349** [reply, marked as answer]
@mislav I tried reinstalling the command line tools and xcode, also restarting. But this does nothing here. We have the same problem un multiple macs of the same type. Its an M2 Max CPU we have a server mac with another M2 type that runs without problems, but the notebooks with the M2 Max seem not to be able to install ruby. 

Whats really interesting is that the MAKE_OPTS fixed the install! Maybe its about the CPU, the Max has more cores than the default one. If you want i can send you more information on the model. This is the command which finally fixed the installation for us:
`RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)" MAKE_OPTS='-j2' rbenv install 3.2.2`

**Reply by Dante1349** [reply]
@mislav the rest of the installation and the usage worked as expected now, thank you very much for your support! If you need more details on the problem im glad to help ;-)

**Reply by mislav** [reply]
Yay, I'm glad it helped. This is the command ruby-build uses to get the number of CPUs: `sysctl -n hw.ncpu` (it's 8 on my M1 MacBook Pro)

Maybe you can try increasing the number from 2 until the maximum number that it starts failing at? Then you can share your findings if you will.

If you happen to know a better number we could use for a default than `sysctl -n hw.ncpu` that would help out-of-the-box experience on machine like yours, I'm open to suggestions.

**Reply by Dante1349** [reply]
for `sysctl -n hw.ncpu` i get `12` on this mac. We have another one now which also fails with these settings, i have to check what kind of cpu it is. I will give you the number as soon as I know

**Reply by Dante1349** [reply]
it seems to work until MAKE_OPTS='-j5' i don't know how this makes sense 😄

**Reply by Dante1349** [reply]
i mean including j5

**Reply by mislav** [reply]
Thanks for sharing the info. It's weird that your machine couldn't handle the compiler parallelism beyond 5 concurrent jobs. A larger number shouldn't giving a modern MacBook a hard time.

Is it possible that your machines are "managed", for instance with software like Jamf? When I last worked for a company, our work machines all had software installed on it that, among other things, monitored processes on the system and prevented execution of those that were deemed to pose a security threat. Sometimes that software would interfere with benign developer tooling.

**Reply by mislav** [reply]
Also, is it possible that your machine does not have sufficient RAM to handle more than 5 concurrent compilation jobs?

I just found some old reports of compilers getting killed by the system after they exhaust available RAM. Not sure if that's really what could be happening here.

**Reply by Dante1349** [reply]
The mac is actually managed via jamf. Don you know if there is a possibility to debug this? I think RAM is not an issue. Its 32gb almost empty when i run the installer.

**Reply by Dante1349** [reply]
Hmm we also have another macbook of the same type where this is not a solution to limit the threads. There is still this "Killed 9" error.

**Reply by Dante1349** [reply]
i have to correct, he probably had the old setup, now it worked for that machine, too with the opts.

**Reply by mislav** [reply]
No, I am at a loss here and I am not proficient enough with system engineering to be able to debug this. If I was in this position I would have contacted the admins who have oversight over Jamf and asked them if they see some outliers on their end. But I can tell you that I haven't heard of reports like yours before. In all the other reports, `killed: 9` issues would have been fixed by either reinstalling the compiler setup (i.e. returning the OS to a more pristine state) or by staying under RAM limits.

**Reply by Dante1349** [reply]
Me neither, I'm a web/app developer. But thank you anyway, this solved my problem at least and i could finish the setup. Thank you very much @mislav

**Comment by rhedwan** [comment]
@Dante1349 
I have tried running the code but it still returns the same error. 
`RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)" MAKE_OPTS='-j2' rbenv install 3.2.2`

<img width="1440" alt="Screenshot 2023-12-02 at 11 47 35 PM" src="https://github.com/rbenv/ruby-build/assets/48237269/827a3bd3-9503-4499-ad1a-65571e41006f">

**Reply by Dante1349** [reply]
@rhedwan I only had this error with the m2 max core, but not on other apple silicon cores. I'm not sure if it's the same issue here.

**Reply by rhedwan** [reply]
@Dante1349 I'm having the problem with my M1.

**Reply by Dante1349** [reply]
@rhedwan on the m1 I could use the installer without any compiler params, I'm not sure what's causing the problem here. Probably not the issue but your macOS seems a little outdated.

**Comment by rhedwan** [comment]
@Dante1349 I have updated my Mac. The issue persists. 
<img width="1440" alt="Screenshot 2023-12-03 at 10 03 11 PM" src="https://github.com/rbenv/ruby-build/assets/48237269/9e70f404-9198-448a-ab80-d7c74cbd6366">

for more context, these are the reasons I'm installing Ruby. 
<img width="1440" alt="Screenshot 2023-12-03 at 10 04 52 PM" src="https://github.com/rbenv/ruby-build/assets/48237269/d9156870-8869-4526-9a3a-a83936a50e8d">

**Reply by monfresh** [reply]
@rhedwan Have you tried installing Ruby 3.1.4 instead of 3.1.2? Is there a particular reason you're using Ruby 3.1.2 for Flutter? You should be able to use the latest Ruby, which is 3.2.2 currently.

These types of issues are usually due to installing older versions of Ruby, or something wrong or outdated with your development setup. Here are some more things you can try:
- Run `flutter doctor -v` for more details.
- Check that Homebrew is happy by running `brew doctor`
- Make sure you're not setting any environment variables that you don't understand in your shell startup file (.zshrc, .zprofile, .zlogin, etc)
- Make sure that `xcode-select -p` points to `/Applications/Xcode.app/Contents/Developer`
- Make sure your Xcode is up to date. I noticed that you are on macOS Sonoma, but your Xcode is 14.2, which is strange. I'm pretty sure that macOS Sonoma requires Xcode 15 or higher.

If you still get stuck, I have a paid product called [Ruby on Mac](https://www.rubyonmac.dev/) that is guaranteed to fix issues with your dev setup and to set up a proper Ruby dev environment. I have several happy customers who used Ruby on Mac to fix their Flutter setup.

**Comment by rhedwan** [comment]
@monfresh  Thank you very much. I appreciate this, it now works. 

<img width="1112" alt="Screenshot 2023-12-04 at 12 04 41 AM" src="https://github.com/rbenv/ruby-build/assets/48237269/66a6ffd1-f2a4-4293-9622-7ae70b0ae889">

**Reply by rhedwan** [reply]
@Dante1349 Thanks a lot. I have been able to fix it.

**Comment by wwwcg** [comment]
Same problem，finally solved with a simple and crude way:
**uninstall brew and then reinstall**

Since my brew was migrated from intel chip, doing this will clean all the files installed in /usr/local, 
and after reinstalling, the directory of brew in the Apple chip moved to /opt/homebrew.
I suspect this is the reason for some of openssl's incompatibilities.
